pkg/workflow/workflow.go

```diff
@@ -0,0 +1,235 @@
+package workflow
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pborman/uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	consts "sigs.k8s.io/azuredisk-csi-driver/pkg/azureconstants"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	k8sRuntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2/klogr"
+	"k8s.io/utils/pointer"
+	diskv1beta1 "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1beta1"
+)
+
+type workflowKey struct{}
+
+type Workflow struct {
+	name           string
+	errSet         *sync.Map
+	startTime      time.Time
+	requestID      string
+	logger         Logger
+	parentWorkflow *Workflow
+	pendingCount   *int32 // pendingCount should always be number of children + 1(itself)
+}
+
+func GetWorkflowFromObj(ctx context.Context, obj k8sRuntime.Object, opts ...Option) (context.Context, Workflow) {
+	w := Workflow{}.applyDefault()
+
+	w.requestID = GetRequestID(obj)
+	startTimeStr := getAnnotationFromRuntimeObject(obj, consts.RequestStartimeKey)
+	if startTimeStr != "" {
+		w.startTime, _ = time.Parse(consts.RequestTimeFormat, startTimeStr)
+	}
+	w.logger = Logger{w.logger.WithValues(append(GetObjectDetails(obj), consts.RequestIDKey, w.requestID)...)}
+
+	for _, opt := range opts {
+		opt(&w)
+	}
+
+	return w.SaveToContext(ctx), w
+}
+
+func GetObjectDetails(obj k8sRuntime.Object) []interface{} {
+	switch target := obj.(type) {
+	case *diskv1beta1.AzVolume:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.PvNameKey, target.Spec.PersistentVolume}
+		}
+	case *diskv1beta1.AzVolumeAttachment:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.NodeNameLabel, target.Spec.NodeName, consts.RoleLabel, target.Spec.RequestedRole}
+		}
+	}
+	return []interface{}{}
+}
+
+type Option func(*Workflow)
+
+func WithDetails(details ...interface{}) Option {
+	return func(w *Workflow) {
+		w.logger = Logger{w.logger.WithValues(details...)}
+	}
+}
+
+// New starts a new child workflow if there is already a workflow saved to the context otherwise a new workflow
+func New(ctx context.Context, opts ...Option) (context.Context, Workflow) {
```

Comment:
We should manage `Workflow` by-reference rather than by-value. Otherwise, the completion won't work correctly in the multi-child case since each child would get new copy of the parent from `GetWorkflowFromContext`.